src/wmtk/Tuple.cpp

```diff
@@ -42,4 +42,14 @@ bool Tuple::same_ids(const wmtk::Tuple& t) const
            std::tie(t.m_local_vid, t.m_local_eid, t.m_local_fid, t.m_global_cid);
 }
 
+bool Tuple::is_null_tuple() const
+{
+    return m_local_vid == -1 && m_local_eid == -1 && m_local_fid == -1 && m_global_cid == -1 && m_hash == -1;
+}
+
+Tuple Tuple::tuple_update_hash(long new_hash) const
```

Comment:
Tuple having a function specifying tuple is unecessary `with_updated_hash(long new_hash)`